### PR TITLE
Allow annotations from current app CRs

### PR DESCRIPTION
**Pull request by tomahawk28**
Toward https://github.com/giantswarm/giantswarm/issues/14677

## Checklist

- [x] Update changelog in CHANGELOG.md.

### CODE REVIEW

File: appresource/new_app_cr_to_update.go [hunk 1]
```diff
@@ -21,6 +21,20 @@ func newAppCRToUpdate(current, desired *v1alpha1.App) *v1alpha1.App {
 	merged := current.DeepCopy()
 
 	merged.Annotations = desired.Annotations
+
+	if allowedAnnotations != nil {
+		for k, v := range current.Annotations {
+			if _, exist := desired.Annotations[k]; exist {
```
Comment (tomahawk28) [inline appresource/new_app_cr_to_update.go:27]: If it's already in the desired annotation, no need to put this again.

File: appresource/resource.go [hunk 2]
```diff
@@ -47,6 +49,17 @@ func New(config Config) (*Resource, error) {
 		name: config.Name,
 	}
 
+	if config.AllowedAnnotations != nil {
+		allowedAnnotation := map[string]bool{}
+		{
+			for _, annotation := range config.AllowedAnnotations {
+				allowedAnnotation[annotation] = true
```
Comment (tomahawk28) [inline appresource/resource.go:56]: I try to get the config by string array, and later construct string map for easier access.

**Comment by tomahawk28** [comment]
Tested in `gauss`
No changes in app CR. 

```
D 12/03 09:17:32 /apis/cluster.x-k8s.io/v1alpha2/namespaces/default/clusters/2tyo9 app.NewUpdatePatch computing App CRs to create  | resource/v2/appresource/create.go:59 | controller=cluster-operator-cluster-controller | event=update | loop=0 | version=62565946
D 12/03 09:17:32 /apis/cluster.x-k8s.io/v1alpha2/namespaces/default/clusters/2tyo9 app.NewUpdatePatch computed 0 App CRs to create | resource/v2/appresource/create.go:67 | controller=cluster-operator-cluster-controller | event=update | loop=0 | version=62565946
D 12/03 09:17:32 /apis/cluster.x-k8s.io/v1alpha2/namespaces/default/clusters/2tyo9 app.NewUpdatePatch computing App CRs to delete | resource/v2/appresource/delete.go:49 | controller=cluster-operator-cluster-controller | event=update | loop=0 | version=62565946
D 12/03 09:17:32 /apis/cluster.x-k8s.io/v1alpha2/namespaces/default/clusters/2tyo9 app.NewUpdatePatch computed 0 App CRs to delete | resource/v2/appresource/delete.go:57 | controller=cluster-operator-cluster-controller | event=update | loop=0 | version=62565946
D 12/03 09:17:32 /apis/cluster.x-k8s.io/v1alpha2/namespaces/default/clusters/2tyo9 app.NewUpdatePatch computing App CRs to update | resource/v2/appresource/update.go:55 | controller=cluster-operator-cluster-controller | event=update | loop=0 | version=62565946
D 12/03 09:17:32 /apis/cluster.x-k8s.io/v1alpha2/namespaces/default/clusters/2tyo9 app.NewUpdatePatch computed 0 App CRs to update | resource/v2/appresource/update.go:66 | controller=cluster-operator-cluster-controller | event=update | loop=0 | version=62565946
```

**Comment by tomahawk28** [comment]
Tested in cluster-operator PR; https://github.com/giantswarm/resource/pull/30

File: CHANGELOG.md [hunk 3]
```diff
@@ -9,6 +9,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Added
+
+- Allowing annotations from current app CR to remain if config have those.
```
Comment (rossf7) [inline CHANGELOG.md:14]:
```suggestion
- Allow annotations from current app CR to remain if config has these. 
```

File: appresource/update.go [hunk 4]
```diff
@@ -20,6 +20,14 @@ func (r *Resource) ApplyUpdateChange(ctx context.Context, obj, updateChange inte
 	for _, appCR := range appCRs {
 		r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("updating App CR %#q in namespace %#q", appCR.Name, appCR.Namespace))
 
+		// Get app CR again to ensure the resource version is correct.
+		currentCR, err := r.g8sClient.ApplicationV1alpha1().Apps(appCR.Namespace).Get(ctx, appCR.Name, metav1.GetOptions{})
+		if err != nil {
+			return microerror.Mask(err)
+		}
```
Comment (rossf7) [inline appresource/update.go:27]: Nice this will make the updates more reliable.

**Review by rossf7 — APPROVED** [review]
LGTM

**Pull request merged**